## Assignment 2: Machine Learning

Yessin Radouane (13428489), Miji Trenkel (13686038)

Please note that due to the fact that some code takes hours to run, we have copied the output and we pasted it in a markdown text box below the code. 

### 1.1 Part One: Classification on Wine Dataset

In this exercise, we ask you to predict wine quality in a multi-class classification setup using a LogisticRegression classifier. In less than 50 words, present the potential advantages and drawbacks of viewing this problem as a multi-class classification task.

In [4]:
import csv
import pandas as pd
import matplotlib.pyplot as plt 
import numpy as np
%matplotlib inline
dataredwine = pd.read_csv('/Users/yessinradouane/Desktop/Uni/ML/Assignment1/winequality/winequality-red.csv', sep=';')
datawhitewine = pd.read_csv('/Users/yessinradouane/Desktop/Uni/ML/Assignment1/winequality/winequality-white.csv', sep=';')

In [2]:
#disable warnings

import warnings
warnings.filterwarnings("ignore")

In [5]:
from collections import Counter
from sklearn.datasets import make_classification
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

features = list(dataredwine.keys())
features.remove('quality')
quality_W = dataredwine['quality']
 
def split_data(data):
    X = data.loc[:, features]
    y = data.loc[:, ['quality']]
 
    X_train, X_test, y_train, y_test = train_test_split(X,y, random_state = 6)
 
    return X_train, X_test, y_train, y_test

def fit_regression (x,y):
    model = LogisticRegression(multi_class = 'multinomial').fit(x,y)
    return model

def regression_summary (model,xtrain,ytrain):

    accuracy = model.score(xtrain,ytrain)
    # print(accuracy)
    return accuracy

In [5]:
# redwine
X_train, X_test, y_train, y_test = split_data(dataredwine)
# X_train.shape
model = fit_regression(X_train, y_train)
print(regression_summary(model, X_train, y_train))

#whitewine
X_train2, X_test2, y_train2, y_test2 = split_data(datawhitewine)
# X_train.shape
model2 = fit_regression(X_train2, y_train2)
print(regression_summary(model2, X_train2, y_train2))

0.5763135946622185
0.46474271712496595


There are two different ways to view the data using logisticRegression: <br> 
Binomial: Target varibles have two possible types <br>
Multi-class: Target variables have three or more possible types <br>
An advantage to viewing this problem as a multi-class classification is probably that estimates will be better calibrated compared to using the default one vs rest (binomial). A disadvantage is that it assumes linearity between dependent and independent variables.


### Part 2 Portuguese Data

The classification goal is to predict if the client will subscribe to a term deposit The boolean label "y". You should train and valid on the train.csv, and test on test.csv and test label.csv.


In [6]:
portuguese_train = pd.read_csv('/Users/yessinradouane/Desktop/Uni/ML/Assignment2/ML_Assignment_ii_A_202223/train.csv', sep=',')
portuguese_test = pd.read_csv('/Users/yessinradouane/Desktop/Uni/ML/Assignment2/ML_Assignment_ii_A_202223/test.csv', sep=',')
portuguese_test_label = pd.read_csv('/Users/yessinradouane/Desktop/Uni/ML/Assignment2/ML_Assignment_ii_A_202223/test_label.csv', sep=',')

In [7]:
portuguese_train.head(10)

ID  age          job   marital  education default  balance housing loan  \
0  13829   29   technician    single   tertiary      no    18254      no   no   
1  22677   26     services    single  secondary      no      512     yes  yes   
2  10541   30   management    single  secondary      no      135      no   no   
3  13689   41   technician   married    unknown      no       30     yes   no   
4  11304   27       admin.    single  secondary      no      321      no  yes   
5  15046   22      student    single  secondary      no      185      no   no   
6  15505   53  blue-collar   married  secondary      no     1777     yes   no   
7  22341   38       admin.  divorced  secondary      no       74     yes   no   
8  18051   59   management   married   tertiary      no     3342      no   no   
9  15826   37    housemaid    single  secondary      no      465     yes   no   

    contact  day month  campaign  pdays  previous poutcome    y  
0  cellular   11   may         2     -1         0  unknown   no  
1   unknown    5   jun         3     -1         0  unknown   no  
2  cellular   14   aug         2     -1         0  unknown   no  
3  cellular   10   jul         1     -1         0  unknown   no  
4   unknown    2   sep         1     -1         0  unknown   no  
5  cellular    8   jul         1     -1         0  unknown  yes  
6  cellular   21   nov         5    154         1  failure  yes  
7  cellular   23   jul         3     -1         0  unknown   no  
8  cellular   18   mar         2    188         6    other  yes  
9  cellular   18   may         3     -1         0  unknown  yes

In [7]:
portuguese_train.dtypes

ID            int64
age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

### 2.3.2: Inspecting feature datatypes
Inspect individual features to ensure they are in the right datatype. Pandas will try to intelligently infer the correct datatype but you still need to inspect the results yourself.
<br>
The two datatype that pandas has classified are int64 and object. Object: The NumPy type for storing strings (and mixed types), and int64: pandas nullable integer type. <br>
The following features have the correct datatype: <br>
**ID, age, balance, day, campaign, pdays, previous:** These are all int64 and are all correctly infered by pandas. As they will always be a number in this given dataframe.<br> 
**job, month, martial, education, poutcome:** Pandas infered them as objects, as in the dataframe they are given as strings, yet later on will we see that they are categorical strings and therefore can be transformed. <br>
**default, housing, loan, contact, y :** These are boolean as they can be viewed as 0 or 1, Pandas infered them as objects as in the dataframe they are given as 'no' or 'yes'. But a better depiction of the datatype would be to transform them into boolean. 






### 2.3.3: One-hot encoding
**Features that contain categorical data :** <br>
Job: admin, blue-collar, entrepreneur, housemaid, management, retired, self-employed, services, student, technocian, unemployed, unknown <br>
Marital: Single, Married, Divorced, unknown <br>
Education: tertiary, secondary, primary, unknown <br>
default: yes, no  <br>
housing: yes, no <br>
loan: yes, no <br> 
contact: cellular, telephone, unknown <br>
month: jan, feb, ..., nov, dec <br>
poutcome: failure,unknown, other, success <br>
y: yes, no (boolean) <br>




In [8]:
portuguese_train['education'].unique()

array(['tertiary', 'secondary', 'unknown', 'primary'], dtype=object)

In [9]:

if 'poutcome' in portuguese_train.keys():
    del portuguese_train["poutcome"]


print("train", portuguese_train.keys())

if 'poutcome' in portuguese_test.keys():

    del portuguese_test["poutcome"]


print("test", portuguese_test.keys())


train Index(['ID', 'age', 'job', 'marital', 'education', 'default', 'balance',
       'housing', 'loan', 'contact', 'day', 'month', 'campaign', 'pdays',
       'previous', 'y'],
      dtype='object')
test Index(['ID', 'age', 'job', 'marital', 'education', 'default', 'balance',
       'housing', 'loan', 'contact', 'day', 'month', 'campaign', 'pdays',
       'previous'],
      dtype='object')


In [10]:
from sklearn.preprocessing import OneHotEncoder

#if the feature is just yes or no / binary then just leave it as that (change to 0,1)

train_dummied = pd.get_dummies(portuguese_train)
test_dummied = pd.get_dummies(portuguese_test)
test_response_dummied = pd.get_dummies(portuguese_test_label)



#remove y_no to get a nice response variable without including the inverse response variable as a explanatory variable.
if 'y_no' in train_dummied.keys():
    
    del train_dummied["y_no"]
    del test_response_dummied['y_no']


train_dummied


ID  age  balance  day  campaign  pdays  previous  job_admin.  \
0      13829   29    18254   11         2     -1         0           0   
1      22677   26      512    5         3     -1         0           0   
2      10541   30      135   14         2     -1         0           0   
3      13689   41       30   10         1     -1         0           0   
4      11304   27      321    2         1     -1         0           1   
...      ...  ...      ...  ...       ...    ...       ...         ...   
12865  14023   47     1167   30         1     87         5           0   
12866  17259   31      111   21         2     93         2           0   
12867  15200   37     1316   18         1    172         2           0   
12868  13775   42      479   28         2     -1         0           0   
12869  20137   24        0   16         1     -1         0           0   

       job_blue-collar  job_entrepreneur  ...  month_feb  month_jan  \
0                    0                 0  ...          0          0   
1                    0                 0  ...          0          0   
2                    0                 0  ...          0          0   
3                    0                 0  ...          0          0   
4                    0                 0  ...          0          0   
...                ...               ...  ...        ...        ...   
12865                0                 0  ...          0          0   
12866                0                 0  ...          0          0   
12867                0                 0  ...          0          0   
12868                0                 0  ...          0          0   
12869                0                 0  ...          0          0   

       month_jul  month_jun  month_mar  month_may  month_nov  month_oct  \
0              0          0          0          1          0          0   
1              0          1          0          0          0          0   
2              0          0          0          0          0          0   
3              1          0          0          0          0          0   
4              0          0          0          0          0          0   
...          ...        ...        ...        ...        ...        ...   
12865          0          0          0          0          0          0   
12866          0          0          0          0          1          0   
12867          0          0          0          0          1          0   
12868          0          0          0          1          0          0   
12869          0          0          0          1          0          0   

       month_sep  y_yes  
0              0      0  
1              0      0  
2              0      0  
3              0      0  
4              1      0  
...          ...    ...  
12865          0      1  
12866          0      1  
12867          0      0  
12868          0      0  
12869          0      0  

[12870 rows x 48 columns]

### 2.4 Models

In [11]:
# scale, classifier into pipeline
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler 
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn import *
from datetime import datetime
from sklearn.metrics import classification_report


def create_pipeline(scalar, classifier):

    start=datetime.now()



    #excluding ID from model
    features = list(train_dummied.keys())
    features.remove('y_yes')
    features.remove('ID')

    #defining data set
    X_trn = train_dummied.loc[:, features]
    y_trn = train_dummied.loc[:, ['y_yes']]
    X_tst = test_dummied.loc[:, features]
    y_tst = test_response_dummied.loc[:, ['y_yes']]


    if classifier == "LinearSVC":

        #defining pipeline
        pipe = make_pipeline(scalar, LinearSVC())

        #get param grid
        c_grid = {"linearsvc__C": np.logspace(-3,3,10)}

        #execute GridSearch
        gridcv = GridSearchCV(pipe, c_grid, cv=5, return_train_score=True)

        #fit data
        gridcv.fit(X_trn, y_trn)

        best_param = gridcv.best_params_

        #y_predicted 
        y_pred = gridcv.predict(X_tst)
        print(classification_report(y_tst, y_pred))


        print(50*"-")
        print(f"{str(scalar)}, {str(classifier)}()")
        print(f"[+] Best parameter: {best_param}\n[+] Test Score: {gridcv.score(X_tst, y_tst)}\n[+] Train Score: {gridcv.score(X_trn, y_trn)}\n[+] Total runtime: {datetime.now()-start}")


    if classifier == "LogisticRegression":

        #defining pipeline
        pipe = make_pipeline(scalar, LogisticRegression())

        #get param grid
        c_grid = {"logisticregression__C": np.logspace(-3,3,10)}

        #execute GridSearch
        gridcv = GridSearchCV(pipe, c_grid, cv=5, return_train_score=True)

        #fit data
        gridcv.fit(X_trn, y_trn)

        best_param = gridcv.best_params_

        #y_predicted 
        y_pred = gridcv.predict(X_tst)
        print(classification_report(y_tst, y_pred))

        print(50*"-")
        print(f"{str(scalar)}, {str(classifier)}()")
        print(f"[+] Best parameter: {best_param}\n[+] Test Score: {gridcv.score(X_tst, y_tst)}\n[+] Train Score: {gridcv.score(X_trn, y_trn)}\n[+] Total runtime: {datetime.now()-start}")


    if classifier == "KNN":
    
        #defining pipeline
        pipe = make_pipeline(scalar, KNeighborsClassifier())

        #get param grid
        n_grid = {"kneighborsclassifier__n_neighbors": range(1,15)}

        #execute GridSearch
        gridcv = GridSearchCV(pipe, n_grid, cv=5, return_train_score=True)

        #fit data
        gridcv.fit(X_trn, y_trn)

        best_param = gridcv.best_params_

        #y_predicted 
        y_pred = gridcv.predict(X_tst)
        print(classification_report(y_tst, y_pred))

        print(50*"-")
        print(f"{str(scalar)}, {str(classifier)}()")
        print(f"[+] Best parameter: {best_param}\n[+] Test Score: {gridcv.score(X_tst, y_tst)}\n[+] Train Score: {gridcv.score(X_trn, y_trn)}\n[+] Total runtime: {datetime.now()-start}")
  

In [ ]:
for scalar in [StandardScaler(), MinMaxScaler()]:
    for classifier in ["LinearSVC", "LogisticRegression", "KNN"]:

        create_pipeline(scalar, classifier)






**0 for y_yes is 'no' and 1 for y_yes is 'yes'**


-------------------------------------------------

StandardScaler(), LinearSVC()

[+] Best parameter: {'linearsvc__C': 0.001}

[+] Test Score: 0.7536704730831973

[+] Train Score: 0.7519036519036519

[+] Total runtime: 0:03:11.637320



              precision    recall  f1-score   support

           0       0.76      0.93      0.84      2969
           1       0.70      0.35      0.46      1322

    accuracy                           0.75      4291
    macro avg       0.73      0.64      0.65      4291
    weighted avg       0.74      0.75      0.72      4291



--------------------------------------------------

StandardScaler(), LogisticRegression()

[+] Best parameter: {'logisticregression__C': 0.021544346900318832}

[+] Test Score: 0.7573992076439059

[+] Train Score: 0.7556332556332557

[+] Total runtime: 0:00:06.185007




              precision    recall  f1-score   support

           0       0.78      0.91      0.84      2969
           1       0.68      0.43      0.52      1322

    accuracy                           0.76      4291
    macro avg       0.73      0.67      0.68      4291
    weighted avg       0.75      0.76      0.74      4291


--------------------------------------------------

StandardScaler(), KNN()

[+] Best parameter: {'kneighborsclassifier__n_neighbors': 13}

[+] Test Score: 0.7615940340247028

[+] Train Score: 0.7791763791763792

[+] Total runtime: 0:07:54.475579



              precision    recall  f1-score   support

           0       0.76      0.94      0.84      2969
           1       0.71      0.34      0.46      1322

    accuracy                           0.75      4291
    macro avg       0.74      0.64      0.65      4291
    weighted avg       0.75      0.75      0.72      4291


--------------------------------------------------

MinMaxScaler(), LinearSVC()

[+] Best parameter: {'linearsvc__C': 0.46415888336127775}

[+] Test Score: 0.7539035189932417

[+] Train Score: 0.7513597513597513

[+] Total runtime: 0:00:43.934720



              precision    recall  f1-score   support

           0       0.77      0.93      0.84      2969
           1       0.70      0.37      0.48      1322

    accuracy                           0.76      4291
    macro avg       0.74      0.65      0.66      4291
    weighted avg       0.75      0.76      0.73      4291



--------------------------------------------------

MinMaxScaler(), LogisticRegression()

[+] Best parameter: {'logisticregression__C': 10.0}

[+] Test Score: 0.7569331158238173

[+] Train Score: 0.7562548562548562

[+] Total runtime: 0:00:21.527807



              precision    recall  f1-score   support

           0       0.77      0.93      0.84      2969
           1       0.70      0.37      0.48      1322

    accuracy                           0.76      4291
    macro avg       0.73      0.65      0.66      4291
    weighted avg       0.75      0.76      0.73      4291


--------------------------------------------------

MinMaxScaler(), KNN()

[+] Best parameter: {'kneighborsclassifier__n_neighbors': 14}

[+] Test Score: 0.7478443253320904

[+] Train Score: 0.768997668997669

[+] Total runtime: 0:07:01.182478



              precision    recall  f1-score   support

           0       0.76      0.93      0.84      2969
           1       0.68      0.34      0.45      1322

    accuracy                           0.75      4291
    macro avg       0.72      0.63      0.64      4291
    weighted avg       0.74      0.75      0.72      4291



**Now we do the same procedure but without scaling**

In [12]:
from sklearn.pipeline import Pipeline

def create_unscaled_pipeline(_classifier):

    start=datetime.now()



    #excluding ID from model
    features = list(train_dummied.keys())
    features.remove('y_yes')
    features.remove('ID')

    #defining data set
    X_trn = train_dummied.loc[:, features]
    y_trn = train_dummied.loc[:, ['y_yes']]
    X_tst = test_dummied.loc[:, features]
    y_tst = test_response_dummied.loc[:, ['y_yes']]


    if _classifier == "LinearSVC":

        #defining pipeline
        pipe = Pipeline(steps=[('model', LinearSVC())])
        # print(pipe.get_params().keys())


        #get param grid
        c_grid = {"model__C": np.logspace(-3,3,10)}

        #execute GridSearch
        gridcv = GridSearchCV(pipe, c_grid, cv=5, return_train_score=True)
        
        #fit data
        gridcv.fit(X_trn, y_trn)

        #best param
        best_param = gridcv.best_params_

        #y_predicted 
        y_pred = gridcv.predict(X_tst)
        print(classification_report(y_tst, y_pred))


        print(50*"-")
        print(f"No scalar, {str(_classifier)}()")
        print(f"[+] Best parameter: {best_param}\n[+] Test Score: {gridcv.score(X_tst, y_tst)}\n[+] Train Score: {gridcv.score(X_tst, y_tst)}\n[+] Total runtime: {datetime.now()-start}")


    if _classifier == "LogisticRegression":
    
        #defining pipeline
        pipe = Pipeline(steps=[('model', LogisticRegression())])
        #print(pipe.get_params().keys())


        #get param grid
        c_grid = {"model__C": np.logspace(-3,3,10)}

        #execute GridSearch
        gridcv = GridSearchCV(pipe, c_grid, cv=5, return_train_score=True)
        
        #fit data
        gridcv.fit(X_trn, y_trn)

        #best param
        best_param = gridcv.best_params_

        #y_predicted 
        y_pred = gridcv.predict(X_tst)
        print(classification_report(y_tst, y_pred))


        print(50*"-")
        print(f"No scalar, {str(_classifier)}()")
        print(f"[+] Best parameter: {best_param}\n[+] Test Score: {gridcv.score(X_tst, y_tst)}\n[+] Train Score: {gridcv.score(X_tst, y_tst)}\n[+] Total runtime: {datetime.now()-start}")

    if _classifier == "KNN":
        
        #defining pipeline
        pipe = Pipeline(steps=[('model', KNeighborsClassifier())])

        #get param grid
        n_grid = {"model__n_neighbors": range(1,15)}

        #execute GridSearch
        gridcv = GridSearchCV(pipe, n_grid, cv=5, return_train_score=True)
        
        #fit data
        gridcv.fit(X_trn, y_trn)

        #best param
        best_param = gridcv.best_params_

        #y_predicted 
        y_pred = gridcv.predict(X_tst)
        print(classification_report(y_tst, y_pred))


        print(50*"-")
        print(f"No scalar, {str(_classifier)}()")
        print(f"[+] Best parameter: {best_param}\n[+] Test Score: {gridcv.score(X_tst, y_tst)}\n[+] Train Score: {gridcv.score(X_tst, y_tst)}\n[+] Total runtime: {datetime.now()-start}")




In [ ]:
for classifier in ["LinearSVC", "LogisticRegression", "KNN"]:
    create_unscaled_pipeline(classifier)




--------------------------------------------------

No scalar, LinearSVC()

[+] Best parameter: {'model__C': 1000.0}

[+] Test Score: 0.6860871591703566

[+] Train Score: 0.6860871591703566

[+] Total runtime: 0:01:46.352529


              precision    recall  f1-score   support

           0       0.74      0.84      0.79      2969
           1       0.49      0.33      0.40      1322

    accuracy                           0.69      4291
    macro avg       0.61      0.59      0.59      4291
    weighted avg       0.66      0.69      0.67      4291



--------------------------------------------------

No scalar, LogisticRegression()

[+] Best parameter: {'model__C': 0.46415888336127775}

[+] Test Score: 0.7359589839198322

[+] Train Score: 0.7359589839198322

[+] Total runtime: 0:00:18.049323


              precision    recall  f1-score   support

           0       0.75      0.92      0.83      2969
           1       0.64      0.32      0.43      1322

    accuracy                           0.74      4291
    macro avg       0.70      0.62      0.63      4291
    weighted avg       0.72      0.74      0.71      4291


--------------------------------------------------

No scalar, KNN()

[+] Best parameter: {'model__n_neighbors': 12}

[+] Test Score: 0.716616173386157

[+] Train Score: 0.716616173386157

[+] Total runtime: 0:06:45.747347


              precision    recall  f1-score   support

           0       0.73      0.93      0.82      2969
           1       0.61      0.23      0.33      1322

    accuracy                           0.72      4291
    macro avg       0.67      0.58      0.58      4291
    weighted avg       0.69      0.72      0.67      4291


**Interpretation linearSVC**

As we can see from the scores, the unscaled linearSVC model (accuracy: 0.68, C = 1000) performs worse than the scaled linearSVC models (StandScale {accuracy: 0.75, C: 0.001}, MinMax {accuracy: 0.75, C: 0.46}). We also see that C increases significantly if we dont scale the data. The C parameter indicates how much it wants to avoid misclassification. Thus, the unscaled data, C = 1000, takes a lot of effort to not misclassify the samples. 


**Interpretation logisticRegression**

As we can see from the scores, the unscaled logisticRegression model (accuracy: 0.74, C = 0.46) performs worse than the scaled logisticRegression models (StandScale {accuracy: 0.76, C: 0.02}, MinMax {accuracy: 0.76, C: 10}). We also see that C increases significantly if we dont scale(StandScale) the data. MinMax scaler does more effort to not misclassify data, but the scores of the scaled versions are higher.

**Interpretation KNN**

As we can see from the scores, the unscaled KNN model (accuracy: 0.72, K = 12) performs worse than the scaled KNN models (StandScale {accuracy: 0.75, K: 13}, MinMax {accuracy: 0.75, K: 14}). Again, scaled data performs best, that comes with a higher K. So more neighbors were needed in the scaled data to improve the score.

Overall we see that the accuracy increases significantly if we scale the data, eventhough it consists of dummy variables. Also notice the low F1 scores, for class 1 (y_yes == yes). it means the models are better at predicting class 'no' than class 'yes' as the f1 score is the harmonic mean of precision and recall.

## 3 Evaluation Metrics

For each of the pipelines, you must report the following classification metrics
1. Accuracy
2. Macro and Micro-Averaged Precision and Recall
3. F1 Score
Additionally, present your observations on what these scores mean for the models under consideration.
These metrics will be discussed at the beginning of Week 5. <br>
**Tip from class: There is a function in Sklearn that generates a report on accuracy macro and mico etc.. into a nice table that makes it easy to read !!!**

These scores are included in the output of the above functions.

# Part b

Initialize a DecisionTreeClassifier model 
(a) Fit the classifier on the data and report the model accuracy

In [13]:
from sklearn.tree import DecisionTreeClassifier

features = list(train_dummied.keys())
#print(features)

X = train_dummied.loc[:, features]
y = train_dummied.loc[:,['y_yes']]
model = DecisionTreeClassifier()
model.fit(X,y)

DecisionTreeClassifier()

(b) Report all evaluation metrics listed in Section 3: <br>
For all models in Assignments II-A and II-B, you must report:
1. Accuracy
2. Macro and Micro-Averaged Precision and Recall
3. F1 Score
Additionally, present your observations on what these scores mean for the models under consideration

In [14]:

features = list(train_dummied.keys())
# print(features)

a = train_dummied.loc[:, features]
y = train_dummied.loc[:,['y_yes']]
X = a.drop(['y_yes'],axis=1)
model_tree = DecisionTreeClassifier()
model_tree.fit(X,y)

['ID', 'age', 'balance', 'day', 'campaign', 'pdays', 'previous', 'job_admin.', 'job_blue-collar', 'job_entrepreneur', 'job_housemaid', 'job_management', 'job_retired', 'job_self-employed', 'job_services', 'job_student', 'job_technician', 'job_unemployed', 'job_unknown', 'marital_divorced', 'marital_married', 'marital_single', 'education_primary', 'education_secondary', 'education_tertiary', 'education_unknown', 'default_no', 'default_yes', 'housing_no', 'housing_yes', 'loan_no', 'loan_yes', 'contact_cellular', 'contact_telephone', 'contact_unknown', 'month_apr', 'month_aug', 'month_dec', 'month_feb', 'month_jan', 'month_jul', 'month_jun', 'month_mar', 'month_may', 'month_nov', 'month_oct', 'month_sep', 'y_yes']


DecisionTreeClassifier()

In [15]:
from sklearn.metrics import classification_report
features = list(test_dummied.keys())

X_tst = test_dummied.loc[:, features]
    
y_pred = model_tree.predict(X_tst)
y_true = test_response_dummied.loc[:, ['y_yes']]


print(classification_report(y_true,y_pred))

              precision    recall  f1-score   support

           0       0.75      0.68      0.71      2969
           1       0.40      0.48      0.44      1322

    accuracy                           0.62      4291
   macro avg       0.57      0.58      0.58      4291
weighted avg       0.64      0.62      0.63      4291



For class 0 the precision is 0.75 this represents the percentage of correct positive predictions, a score of 75% is good but could be better. For class 1 the precision drops to 40%, but the weighted average precision is 0.64, which is okay but it is low due to the fact that we have not selected/tunned any hyperparameters. The (weighted average) recall is 0.62 which is also not as high as you would want from your model. 
The model is better at predicting class 0 over class 1, the f1-score for class 0 is 0.71 and for class 1 it is 0.44, the closer to 1 the better the model. We can see that it is alot better at predicting class 0 and that might be due to the support as the data is slightly imbalanced with class 0 having a larger support and therefore the model is able to predict it more accuracy. 

**(c) Answer the question : Are these models resilient to overfitting when model hyperparameters
have not been carefully selected? Supplement your explanation with suitable figures/tables if
necessary. Please limit your explanation to a maximum of 50 words.**

Decision trees are used as they are easily visualised and understood by non-experts, as well as invariant to scaling of data, yet one of their disadvantages is that they tend to overfit and provide poor generalisation compared to other models. The  hyperparameters we look at with decision trees are max depth, max features, max leaf nodes etc.., with these we can tune the model to prevent overfitting. If the hyperparameters are not selected carefully it will overfit as the model is not resilient to overfitting. 

Look to the figure below, the same regression but one with max depth of 2 and one with max depth of 50. We can see the greater the max depth the more complex the decision tree become, and therefore it's higher overfitted to the training data. 
To aid in this we should use cross-validation and grid-search to find the optimal hyperparameters, which allows a trade-off between under and over fitted decision tree. 

**(d) Perform a grid search on the decision tree model parameters [‘max depth’,
‘max features’] to evaluate the optimal values for these parameters. Report model hyperparameters for the best classifier model.**

In [16]:
from sklearn.model_selection import GridSearchCV
params =  { 'max_features':range(1,15),'max_depth':range(1,15)}
grid = GridSearchCV(estimator=model_tree,param_grid=params,cv=10,n_jobs=1,verbose=2)
grid.fit(X, y)
bestparams = grid.best_params_
bestscore = grid.best_score_
bestestimator= grid.best_estimator_

Fitting 10 folds for each of 196 candidates, totalling 1960 fits
[CV] END ........................max_depth=1, max_features=1; total time=   0.1s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.1s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=1; total time=   0.0s
[CV] END ........................max_depth=1, max_features=2; total time=   0.0s
[CV] END ........................max_depth=1

In [20]:
print(bestparams)
print(bestscore)
print(bestestimator)

{'max_depth': 7, 'max_features': 7}
0.15530643262872398
DecisionTreeRegressor(max_depth=7, max_features=7)


**Initialize a RandomForestClassifier model and perform the following tasks:
(a) Fit the classifier on the data and report the model accuracy.**

In [21]:
from sklearn.ensemble import RandomForestClassifier
features = list(train_dummied.keys())
a = train_dummied.loc[:, features]
y = train_dummied.loc[:,['y_yes']]
X = a.drop(['y_yes'],axis=1)

model_randomforest = RandomForestClassifier()
model_randomforest.fit(X,y)

RandomForestClassifier()

(b) Report all evaluation metrics listed in Section 3.
1. Accuracy
2. Macro and Micro-Averaged Precision and Recall
3. F1 Score
Additionally, present your observations on what these scores mean for the models under consideration

In [22]:
from sklearn.metrics import classification_report
features = list(test_dummied.keys())

X_tst = test_dummied.loc[:, features]
    
y_pred = model_randomforest.predict(X_tst)
y_true = test_response_dummied.loc[:, ['y_yes']]


print(classification_report(y_true,y_pred))

              precision    recall  f1-score   support

           0       0.80      0.91      0.85      2969
           1       0.71      0.50      0.59      1322

    accuracy                           0.78      4291
   macro avg       0.76      0.70      0.72      4291
weighted avg       0.77      0.78      0.77      4291



Compared to the decision tree the random forest is a better suited model for our data, in every component it scored a higher score. Yet we see similarities in class 0 having higher score compared to class 1, due to the support as the data is slighly imbalanced between the classes. Class 0 having a larger support and therefore being able to predict better, with a recall of 0.91 compared to class 1 with a recall of 0.50, as well as class 0 has an f1-score of 0.85 and class 1 a f1-score of 0.59. We would expect the random forest to be a 'better' model as it has more tree which help correct each other. 

**(c) Answer the following questions:<br>
i. How are decision tree classifiers different from random forests on a structural level? (max.
50 words)<br>
ii. Where would you choose decision trees over random forests and vice-versa? Demonstrate
this using an appropriate example from your data. (max. 50 words)<br>
iii. Is accuracy an appropriate evaluation metric for this classification task? if yes, in what
kind of data may it not be a good metric? Justify your answer in less than 20 words.**

i. Decision trees keep spliting until its leaves, it does this by learning if/else questions (decisions) then it goes down to another path and keeps going until it stops and forms a leaf. Meanwhile random forests are a collection of decision trees. Where all trees in the random forest are different: because of boostrap sampling and selection of features in each node. 

ii. It depends on the goal, there is a trade off between interpretability and accuracy when choosing between trees and forests. If the dataset given  goal was to interpret the results in a comprehensible manner or to give reasoning behind predictions I would choose trees, as they train quickly, give results in a very easy communicative way. If it is about data understanding tree are the way to go, but if it is about predicting the data then I would choose forests; as the accuracy is higher, produces more reliable results, is more robust, yet it does take more computational power and time but yield more accurate results, therefore given this data set I would select random forest.

iii. Here accuracy is an appropriate evaluation as we are using multiple features and seeing if we can accurately predict if or not a client will subscribe to a term deposit. It's a good metric here as we have the following data availble, train, and test. Therefore we can check how 'well' our model is able to predict. An example of data where accuracy is not an appropriate evaluation metric is imbalanced data; distribution of examples across the known classes is biased, eg. if an email is spam or not. The issue is the minority is often ignored if the data is highly imbalance and machine learning algorithms will favor the the larger class.

(d) Perform a grid search on the random forest model parameters [‘max depth’,
‘max features’ ‘n estimators’] to evaluate the optimal values for these parameters. Report model parameters for your best classifier model.
In no more than 50 words, explain your observations and your assessment of the classifier. You
can use a text box (i.e., Markdown Cell) in Jupyter to write down your analysis

In [23]:
from sklearn.model_selection import GridSearchCV
params =  { 'max_features':['auto', 'sqrt'],'max_depth':range(1,5), 'n_estimators':range(100,150)}
grid = GridSearchCV(estimator=model_randomforest,param_grid=params,cv=10)
grid.fit(X, y)
bestparams = grid.best_params_
bestscore = grid.best_score_
bestestimator= grid.best_estimator_

In [24]:
print(bestparams)
print(bestscore)
print(bestestimator)

{'max_depth': 4, 'max_features': 'auto', 'n_estimators': 130}
0.7401709401709402
RandomForestClassifier(max_depth=4, n_estimators=130)


**OUTPUT OF ABOVE CODE**

{'max_depth': 4, 

'max_features': 'auto', 

'n_estimators': 130}

0.7401709401709402

RandomForestClassifier(max_depth=4, n_estimators=130)

In [ ]:
features_grid = list(train_dummied.keys())
a_grid = train_dummied.loc[:, features_grid]
y_grid = train_dummied.loc[:,['y_yes']]
X_grid = a_grid.drop(['y_yes'],axis=1)
model_randomforest_grid = RandomForestClassifier(max_depth=4, max_features= 'auto', n_estimators=148)
model_randomforest_grid.fit(X_grid,y_grid)

In [ ]:
features_grid = list(test_dummied.keys())

X_tst_grid = test_dummied.loc[:, features_grid]
    
y_pred = model_randomforest_grid.predict(X_tst_grid)
y_true = test_response_dummied.loc[:, ['y_yes']]

print(classification_report(y_true,y_pred))

Sadly the above code took so long to run that we never got the output.